### ISSUE TITLE
Any dynamic route renders twice in browser

### ISSUE THREAD

**Issue by kyteague** [zeit/next.js#12010]
# Bug report

## Describe the bug

Any dynamic route is rendered twice in the browser. The entire `<App />` seems to be rendered twice.

## To Reproduce

1. Create route [pid].js in `pages`
2. Add code
```javascript
const Page = () => {
  console.log('rendered')
  return <div />
}

export default Page
```
3. `npm run dev`
3. Open browser to `localhost:3000/anypage`
4. See 'rendered' printed twice

## Expected behavior

Dynamic routes should only render once in the browser like any other page.

## System information

macOS 10.15.4
Next.js 9.3.4
Chrome 80.0.3987.163 + Firefox 74.0

**Comment by mattcarlotta** [comment]
I am able to reproduce this issue in v9.3.5 as well. Doesn't matter if it's a pure function or a class component, the `return` statement is executed twice in a dynamic page and executed once in the static home page.

Link to [example repo](https://github.com/mattcarlotta/next-dynamic-page-bug)
or
Clone this repo: `git@github.com:mattcarlotta/next-dynamic-page-bug.git`
Install dependencies: `yarn`
Run dev: `yarn dev`
Go to http://localhost:3000/1.

**Comment by ijjk** [comment]
Hi, this is the expected behavior for automatically statically optimized dynamic pages. 

When a dynamic page does not leverage `getServerSideProps`, `getStaticProps`, or `getInitialProps` it is auto prerendered at build time and since the dynamic page's values can't be known during this stage the initial HTML doesn't isn't rendered with them so during the first render on the client we can't provide these values either to make sure to match the initial HTML and not break hydration. 

After hydration has occurred we parse the values from the URL on the client and trigger another render to allow the parsed values to be used now that hydration has occurred.

**Comment by kyteague** [comment]
@ijjk Thanks. Makes total sense.

**Comment by felipenmoura** [comment]
Hi.
Ok, but what's the solution then?

I even tried using the getStaticProps with empty props:

```
export async function getStaticProps (context) {
  return {
    props: { }
  };
}
```

It still renders twice (what makes lighthouse say it is taking way longer for the first meaningful paint than it actually is).
Is there anything else that may be triggering this full page re-rendering?

**Comment by Lazercat** [comment]
A cheap work around - 
Its possible to persist a state boolean value in a store such as redux you can have the first render set it via your `[dynamicPage].tsx`, then have vital functions look to that boolean in your store first before setting them...  not the prettiest thing ever, but maybe it'll make your day better :) i have analytics firing twice from a child and this helps me solve it in the children comps that call those analytics.   bandaid approach but works.. 


On your [dynamic-page].tsx...
```
const { prerenderCompleted, setPrerenderCompleted } = useStore('redux or whatever you use - we use react hooks');

 useEffect(() => {
    console.log('dynamic page mounted');
    if (prerenderCompleted) {
      console.log('i prerendered yay');
      
      //calls unmount if second mount unmounts.. clean up value
      return () => {
        setPrerenderCompleted(false);
        console.log('post prerender unmount');
      };
    }
    setPrerenderCompleted(true);
    //TIP: in useEffect's, return function fires on unmount..
    return () => console.log('dynamic page unmounted');
  }, []);
```
  
  then, in your child component creating your headaches :) 
 ```
 const { prerenderCompleted } = useStore('redux or whatever you use - we use react hooks');
 if(prerenderCompleted){  doFoo(); }
 ```

**Comment by rickchristie** [comment]
I just want to use query string and found out that the page is rendered twice whenever page is accessed with query string. **Even if the page is not using `useRouter`, etc.**

Follow the guide at https://nextjs.org/docs/api-reference/data-fetching/getInitialProps - using typescript:

```js
import { NextPage } from 'next'

interface Props {
  userAgent?: string;
}

const Page: NextPage<Props> = ({ userAgent }) => (
  <main>Your user agent: {userAgent}</main>
)

Page.getInitialProps = async ({ req }) => {
  const userAgent = req ? req.headers['user-agent'] : navigator.userAgent
  return { userAgent }
}

export default Page
```

Setting a `getInitialProps` makes the render happen once.

**Comment by davidkhierl** [comment]
> I just want to use query string and found out that the page is rendered twice whenever page is accessed with query string. **Even if the page is not using `useRouter`, etc.**
> 
> For static websites, we can use `React.memo` to prevent the second re-render, if the props hasn't changed.
> 
> See: https://reactjs.org/docs/react-api.html#reactmemo
> 
> Hope this helps someone!

this happens to me as well, adding query parameter in the url renders the page twice

**Comment by vf1** [comment]
@rickchristie, do you use memo on page component or entire app (_app.tsx)? I would like to wrap _app, but I not sure that there are no any side effects.

**Comment by dohomi** [comment]
@ijjk thanks for the explanation. Would it not be ideal if NextRouter would split the hook into two hooks, one which renders twice for query params like `useQuery` and one for `useRouter` which doesn't render twice?

**Comment by ijjk** [comment]
@dohomi more granular hooks for handling the query updating would need to be added carefully to prevent breaking change although feel free to open a feature request for this [here](https://github.com/vercel/next.js/discussions/new?category=ideas)

**Comment by dohomi** [comment]
@ijjk ok I understand that totally, that needs to be part of a major upgrade. 
What I think might be good at the moment at https://nextjs.org/docs/advanced-features/custom-app, to hightlight the render mechanism of _app:
```js
import App from 'next/app'
import { useMemo } from 'react'

function MyApp({ Component, pageProps }) {
  // memoize page component render for SSG pages
  return useMemo(
    () => <Component {...pageProps} />, 
    [pageProps])
}
```

**Comment by EmirBoyaci** [comment]
> @dohomi more granular hooks for handling the query updating would need to be added carefully to prevent breaking change although feel free to open a feature request for this [here](https://github.com/vercel/next.js/discussions/new?category=ideas)



> @ijjk ok I understand that totally, that needs to be part of a major upgrade. What I think might be good at the moment at https://nextjs.org/docs/advanced-features/custom-app, to highlight the render mechanism of _app:
> 
> ```js
> import App from 'next/app'
> import { useMemo } from 'react'
> 
> function MyApp({ Component, pageProps }) {
>   // memoize page component render for SSG pages
>   return useMemo(
>     () => <Component {...pageProps} />, 
>     [pageProps]
>   )
> }
> ```
> 
> Also it might be good to highlight this render behaviour as part of https://nextjs.org/docs/advanced-features/custom-app#caveats

Is it still continuing in Next.js v12.0.7?

**Comment by ivan-kleshnin** [comment]
@dohomi

```ts
return useMemo(
  () => <Component {...pageProps} />, 
  [pageProps]
)
```

shouldn't `Component` be a `useMemo` dependency?

```
[Component, pageProps]
```

**Comment by ivan-kleshnin** [comment]
@dohomi

```tsx
return useMemo(
  () => <Component {...pageProps} />, 
  [pageProps]
)
```

will reuse memoized `Component` if the same page with different id is accessed. Like `blog/1 -> blog/2` will mistakenly return `blog/1` under `blog/2`.

It should rather be something like:

```tsx
return useMemo(
  () => <Component {...pageProps} />, 
  [router.asPath]
)
```

@EmirBoyaci yes, nothing was changed in that regard, unless I've missed something in changelog

**Comment by dohomi** [comment]
@ivan-kleshnin my example is shown by the fact that the pageProps are propagated through a CMS and thus their payload is different from page to page. Thats why for my use case the example is right.

**Comment by EmirBoyaci** [comment]
@ivan-kleshnin i'm getting the twice rendering problem with url search params. If I've a page which includes url search params, it automatically renders twice in any page. For example: abc.com (render only once), abc.com/?a=b (renders twice). I tried with using useMemo but it didn't worked for me. Do you have any idea?

Thanks,
Emir

**Comment by ivan-kleshnin** [comment]
@EmirBoyaci

That's a well-known case: https://github.com/vercel/next.js/issues/12010

> When a dynamic page does not leverage getServerSideProps, getStaticProps, or getInitialProps it is auto prerendered at build time and since the dynamic page's values can't be known during this stage the initial HTML doesn't isn't rendered with them so during the first render on the client we can't provide these values either to make sure to match the initial HTML and not break hydration.

You don't want to prevent it, otherwise React will complain about [mismatch](https://reactjs.org/docs/react-dom.html)

**Comment by EmirBoyaci** [comment]
> @EmirBoyaci
> 
> 
> 
> That's a well-known case: https://github.com/vercel/next.js/issues/12010
> 
> 
> 
> > When a dynamic page does not leverage getServerSideProps, getStaticProps, or getInitialProps it is auto prerendered at build time and since the dynamic page's values can't be known during this stage the initial HTML doesn't isn't rendered with them so during the first render on the client we can't provide these values either to make sure to match the initial HTML and not break hydration.
> 
> 
> 
> You don't want to prevent it, otherwise React will complain about DOM [mismatch](https://reactjs.org/docs/react-dom.html).

I got it but the problem for me is below:
I don't complain why it's renders twice but I complain about that, when it renders twice, it sends requests that I've sending in that page twice. I've another page which doesn't have url search params but it has **getStaticProps** and the page doesn't rendered twice. I'm getting this thing while only using url search params (I guess?)

**Comment by ivan-kleshnin** [comment]
If I understand you correctly, you'd like to unify what's going on on different pages WITH and WITHOUT query params and also in SERVER and in BROWSER. That's what our team uses:

```tsx
export function WaitForRouter({children, page = false, hidden = false} : WaitForRouterProps) : JSX.Element {
  const router = useRouter()
  const [ready, setReady] = React.useState(false)

  React.useEffect(() => {
    setReady(true)
  }, [])

  // `ready` check is necessary for empty query cases where
  // !router.isReady on BE and
  // router.isReady immediately on FE
  if (ready && router.isReady)
    return <>{children}</>

  if (!hidden) {
    return <Loader page={page}/>
  }

  return <></>
}
```

And then, in Layouts where you don't expose `getServerSideProps` or `getStaticProps`:

```tsx
<WaitForRouter>
  <Header/>
  <WidthHolder>
    <Controller/>
  </WidthHolder>
</WaitForRouter>
```

**Comment by madou** [comment]
@ijjk What seems to be happening though is a double _mount_, not just a render. So effects are running when they shouldn't, for example this code:

```
// pages/app/[id].js
function Route() {
  useEffect(() => {
    console.log("I should run once, but actually run twice");
  }, []);

  return <div>hi</div>;
}
```

This seems like a bug. Can I get your thoughts in the matter please?

**Comment by ivan-kleshnin** [comment]
@madou double mount is always caused by ancestor components. The above code is not enough to answer.

**Comment by madou** [comment]
Fair. I'll see if I can put together a separate reproduction and raise another issue if it turns out to be something 🙂.

**Comment by terry-fei** [comment]
```js
function Route() {
  const router = useRouter();

  useEffect(() => {
    if (!router.isReady) return;
    console.log("will get right query here");
  }, [router.isReady]);

  return <div>hi</div>;
}
```

**Comment by Joshua-Enrico** [comment]
> When using dynamic route I got that problem, that solution mitigated the double execution of a query, I will be waiting for more news about this, thanks 
![image](https://user-images.githubusercontent.com/77980741/150036561-bd20f4c7-5396-4601-8aa8-a15cf254ac8b.png)

**Comment by ivan-kleshnin** [comment]
@terry-fei this code snippet (mentioned in docs) allows to work with query only inside an effect.
In many cases it's not sufficient.
